base_layer/core/src/proof_of_work/lwma_diff.rs

```diff
@@ -0,0 +1,256 @@
+        j += 1;
+        weighted_times += solvetime * j;
+
+        avg_target += target_difficulties[i].as_u64();
+    }
+    let avg = k / weighted_times as f64;
+    ((avg * avg_target as f64 / n) as u64).into()
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_difficulty_adjustment() {
+        let mut dif = LWMA::new();
+        let mut counter = 60;
+        dif.add(counter, 100.into());
+        counter += 60;
+        dif.add(counter, 100.into());
+        let mut curr_diff = dif.get_difficulty();
+        let mut prev_diff = dif.get_difficulty();
+
+        for _i in 0..150 {
+            prev_diff = curr_diff;
+            counter += 60;
+            dif.add(counter, prev_diff);
+            curr_diff = dif.get_difficulty();
+        }
+
+        assert_eq!(curr_diff == prev_diff, true);
+        // The difficulty stayed the same, lets increase the difficulty
+
+        for _i in 0..60 {
+            prev_diff = curr_diff;
+            counter += 30;
+            dif.add(counter, prev_diff);
+            curr_diff = dif.get_difficulty();
+        }
+
+        assert_eq!(curr_diff > prev_diff, true);
+
+        // difficulty increased, lets decrease diff
+
+        for _i in 0..60 {
+            prev_diff = curr_diff;
+            counter += 90;
+            dif.add(counter, prev_diff);
+            curr_diff = dif.get_difficulty();
+        }
+
+        assert_eq!(curr_diff < prev_diff, true);
+        assert_eq!(dif.target_difficulties.len(), 151);
+        assert_eq!(dif.timestamps.len(), 151);
+    }
+
+    #[test]
+    fn test_zero_len() {
+        let mut dif = LWMA::new();
+        assert_eq!(dif.get_difficulty(), Difficulty::min());
+    }
+
+    #[test]
+    fn test_negative_solve_times() {
+        let mut dif = LWMA::new();
+        let mut counter = 60;
+        dif.add(counter, 100.into());
+        counter += 60;
+        dif.add(counter, 100.into());
+        let mut curr_diff = dif.get_difficulty();
+        let mut prev_diff = dif.get_difficulty();
+
+        // Lets create a history and populate the vecs
+        for _i in 0..150 {
+            prev_diff = curr_diff;
+            counter += 60;
+            dif.add(counter, prev_diff);
+            curr_diff = dif.get_difficulty();
+        }
+
+        // lets create chaose by having 60 blocks as negative solve times. This should never be allowed in practive by
+        // having checks on the block times.
+
+        for _i in 0..60 {
+            prev_diff = curr_diff;
+            counter -= 1; // Only choosing -1 here since we are testing negative solve times and we cannot have 0 time
+            dif.add(counter, prev_diff);
+            curr_diff = dif.get_difficulty();
+            // Algo should handle this as 1sec solve time thus increase the difficulty constantly
+            assert_eq!(curr_diff > prev_diff, true);
+        }
+    }
+
+    #[test]
+    fn test_calculations() {
```

Comment:
Could you add a comment with the test data here, so it's easy to see the expected results?